src/Esprima/ArrayList.cs

```diff
@@ -120,7 +150,62 @@ public int Count
         }
     }
 
-    internal void AddRange<TSource>(ArrayList<TSource> list) where TSource : T
+    public T this[int index]
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        get
+        {
+            AssertUnchanged();
+
+            // Following trick can reduce the range check by one
+            if ((uint) index < (uint) _count)
```

Comment:
To avoid check I think you first need to create a local array reference, foreach does the same.